src/mpi/coll/allreduce/allreduce_intra_recursive_multiplying.c

```diff
@@ -0,0 +1,233 @@
+    MPIR_Datatype_get_extent_macro(datatype, extent);
+    extent = MPL_MAX(extent, true_extent);
+
+    MPIR_CHKLMEM_MALLOC(tmp_buf, void *, (k - 1) * count * extent, mpi_errno,
+                        "temporary buffer", MPL_MEM_BUFFER); // cut to half
+    MPIR_CHKLMEM_MALLOC(reqs, MPIR_Request **, (2 * (k - 1) * sizeof(MPIR_Request *)), mpi_errno,
+                        "reqs", MPL_MEM_BUFFER);
+
+    /* adjust for potential negative lower bound in datatype */
+    tmp_buf = (void *) ((char *) tmp_buf - true_lb);
+
+    /* copy local data into recvbuf */
+    if (sendbuf != MPI_IN_PLACE) {
+        mpi_errno = MPIR_Localcopy(sendbuf, count, datatype, recvbuf, count, datatype);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    /* get nearest power-of-two less than or equal to comm_size */
+    p = (int) (log(comm_size) / log(k));
+    pofk = (int) lround(pow(k, p));
+
+    /* Compute group-specific information */
+    comm_size_div_pofk = comm_size / pofk;
+    comm_size_mod_pofk = comm_size % pofk;
+
+    if (rank > (comm_size_div_pofk + 1) * (comm_size_mod_pofk)) {
+        group = (comm_size_mod_pofk) + (rank - (comm_size_div_pofk + 1) * comm_size_mod_pofk) / comm_size_div_pofk;
+    } else {
+        group = rank / (comm_size_div_pofk + 1);
+    }
+
+    group_size = (comm_size_div_pofk) + (group < (comm_size_mod_pofk) ? 1 : 0);
+    group_leader = group * (comm_size_div_pofk) + MPL_MIN(group, comm_size_mod_pofk) + group_size - 1;
+
+    /* PRE-STAGE: Followers should send data to leaders and leaders 
+     * should receive data from followers */
+    if (rank == group_leader) {
+        /* This is leader so receive data */
+        for (i = 0; i < group_size - 1; i++) {
+            dst = rank - group_size + i + 1;
+            mpi_errno = MPIC_Irecv(((char *)tmp_buf) + i * count * extent, count,
+                                  datatype, dst,
+                                  MPIR_ALLREDUCE_TAG, comm_ptr, &reqs[num_reqs++]);
+            if (mpi_errno) {
+                MPIR_ERR_POP(mpi_errno);
+            }
+        }
+        
+        MPIC_Waitall(num_reqs, reqs, MPI_STATUSES_IGNORE);
+
+        for (i = group_size - 2; i >= 0; i--) {
+            /* Perform local reduction */
+            mpi_errno = MPIR_Reduce_local(((char *)tmp_buf) + i * count * extent, recvbuf, count, datatype, op);
+            MPIR_ERR_CHECK(mpi_errno);  
+        }
+
+        /* Virtual rank should be used for leader for recursive operations */
+        virt_rank = group;
+    } else {
+        /* This is follower so send data */
+        mpi_errno = MPIC_Send(recvbuf, count,
+                              datatype, group_leader, MPIR_ALLREDUCE_TAG, comm_ptr, errflag);
+        MPIR_ERR_CHECK(mpi_errno);
+        /* Virtual rank should not be used */
+        virt_rank = -1;
+    }
+
+    /*MAIN-STAGE*/
+    if (virt_rank != -1) {
+        radix_size = 1;
+        next_radix_size = k;
+
```

Comment:
Add a comment explaining your strategy in the main stage. The comment don't need be in much detail, but should let an uninitialized (but seasoned) maintainer understand or guess what you are trying to do.